src/OrleansAWSUtils/Storage/DynamoDBStorage.cs

```diff
@@ -0,0 +1,430 @@
+﻿using Amazon.DynamoDBv2;
+using Amazon.DynamoDBv2.Model;
+using Amazon.Runtime;
+using Orleans;
+using Orleans.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OrleansAWSUtils.Storage
+{
+    internal class DynamoDBStorage
+    {
+        private const string AccessKeyPropertyName = "AccessKey";
+        private const string SecretKeyPropertyName = "SecretKey";
+        private const string ServicePropertyName = "Service";
+        private const string ReadCapacityUnitsPropertyName = "ReadCapacityUnits";
+        private const string WriteCapacityUnitsPropertyName = "WriteCapacityUnits";
+
+        private string accessKey;
+        protected string secretKey;
+        private string service;
+        private int readCapacityUnits = 10;
+        private int writeCapacityUnits = 5;
+        private AmazonDynamoDBClient ddbClient;
+        private Logger Logger;
+
+        public DynamoDBStorage(string dataConnectionString, Logger logger = null)
+        {
+            ParseDataConnectionString(dataConnectionString);
+            Logger = logger ?? LogManager.GetLogger($"DynamoDBStorage", LoggerType.Runtime);
+            CreateClient();
+        }
+
+        public async Task InitializeTable(string tableName, List<KeySchemaElement> keys, List<AttributeDefinition> attributes)
+        {
+            try
+            {
+                if (await GetTableDescription(tableName) == null)
+                    await CreateTable(tableName, keys, attributes);
+            }
+            catch (Exception exc)
+            {
+                Logger.Error(ErrorCode.StorageProviderBase, $"Could not initialize connection to storage table {tableName}", exc);
+                throw;
+            }
+        }
+
+        #region Table Management Operations
+
+        private void ParseDataConnectionString(string dataConnectionString)
+        {
+            var parameters = dataConnectionString.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
+
+            var serviceConfig = parameters.Where(p => p.Contains(ServicePropertyName)).FirstOrDefault();
+            if (!string.IsNullOrWhiteSpace(serviceConfig))
+            {
+                var value = serviceConfig.Split(new[] { '=' }, StringSplitOptions.RemoveEmptyEntries);
+                if (value.Length == 2 && !string.IsNullOrWhiteSpace(value[1]))
+                    service = value[1];
+            }
```

Comment:
this block, including the line above it should be refactored as a method. now it's just duplicated 4 times